### PR TITLE
BLOCKCHAIN-164 - Congress & Senate should be pallet-democracy BlacklistOrigin

### PR DESCRIPTION
**Pull request by MichalLustykNeti**
Substrate PR: https://github.com/liberland/liberland_substrate/pull/329

image

### CODE REVIEW

File: src/api/nodeRpcCall.js [hunk 1]
```diff
@@ -1912,6 +1912,16 @@ const congressSendTreasuryLld = async (transferToAddress, transferAmount, wallet
   return await submitExtrinsic( extrinsic, walletAddress);
 }
 
+const createBlacklistMotion = async (proposalHash, referendumIndex, walletAddress) => {
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:1915]:
```suggestion
const congressDemocracyBlacklist = async (proposalHash, referendumIndex, walletAddress) => {
```

The naming convention for creating motions is `congress[SomeAction]`.

File: src/components/Voting/Referendum/Items/ProposalItem/index.js [hunk 2]
```diff
@@ -15,6 +23,30 @@ const endorseButton = (buttonEndorseCallback, userDidEndorse, referendumInfo) =>
...
+const blacklistButton = (proposalHash) => {
+  const dispatch = useDispatch();
+  const members = useSelector(congressSelectors.members);
+  const user = useSelector(blockchainSelectors.userWalletAddressSelector);
+
+  useEffect(() => {
+    dispatch(congressActions.getMembers.call());
+  }, [dispatch]);
+
+  const userIsMember = members.map((m) => m.toString()).includes(user);
```
Comment (kacperzuk-neti) [inline src/components/Voting/Referendum/Items/ProposalItem/index.js:35]: Leave a comment to refactor this after #143 is merged

File: src/components/Voting/Referendum/Items/ReferendumItem/index.js [hunk 3]
```diff
@@ -26,16 +50,67 @@ const alreadyVotedButton = (alreadyVoted) => (alreadyVoted === 'Aye'
...
+const blacklistButton = (proposalHash, referendumIndex) => {
+  const dispatch = useDispatch();
+  const members = useSelector(congressSelectors.members);
+  const user = useSelector(blockchainSelectors.userWalletAddressSelector);
+
+  useEffect(() => {
+    dispatch(congressActions.getMembers.call());
+  }, [dispatch]);
+
+  const userIsMember = members.map((m) => m.toString()).includes(user);
```
Comment (kacperzuk-neti) [inline src/components/Voting/Referendum/Items/ReferendumItem/index.js:68]: Leave a comment to refactor this after #143 is merged.

File: src/components/Voting/Referendum/Items/ReferendumItem/index.js [hunk 4]
```diff
@@ -26,16 +50,67 @@ const alreadyVotedButton = (alreadyVoted) => (alreadyVoted === 'Aye'
...
+       * 100}%`
+    : '0%';
   return (
     <Card
       title={name}
       className={styles.referendumItemContainer}
     >
       <div>
+        <div className={styles.rowEnd}>
+          {blacklistButton(hash, referendumIndex)}
```
Comment (kacperzuk-neti) [inline src/components/Voting/Referendum/Items/ReferendumItem/index.js:112]: This should be `<BlacklistButton hash={hash} referendumIndex={referendumIndex} />` or we risk running into hook call order issues.

File: src/components/Voting/Referendum/Items/ProposalItem/index.js [hunk 5]
```diff
@@ -37,14 +69,28 @@ function BoundedCall({ call }) {
...
+  proposalIndex,
 }) {
   return (
     <Card
       title={name}
       className={styles.cardProposalsSection}
     >
       <div>
+        <div className={styles.rowEnd}>
+          {blacklistButton(
```
Comment (kacperzuk-neti) [inline src/components/Voting/Referendum/Items/ProposalItem/index.js:89]: This should be `<BlacklistButton hash={hash} />` or we risk running into hook call order issues.

File: src/redux/actions/congress.js [hunk 6]
```diff
@@ -19,6 +19,7 @@ export const {
   unapproveTreasurySpend,
   voteAtMotions,
   congressProposeRepealLegislation,
+  createBlacklistMotion,
```
Comment (kacperzuk-neti) [inline src/redux/actions/congress.js:22]: Inconsistent naming convention - make it `congressDemocracyBlacklist`.

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

File: src/components/Voting/Referendum/Items/ProposalItem/index.js (hunk 2 above)
Comment (MichalLustykNeti) [reply src/components/Voting/Referendum/Items/ProposalItem/index.js:36]: Addrd comment and created task on jira https://liberland.atlassian.net/jira/software/projects/BLOCKCHAIN/boards/2?selectedIssue=BLOCKCHAIN-222

**Comment by DorianSternVukotic** [comment]
New task in Jira will be created to handle multiple people having the same button, which will create multiple motions. Same thing for legislation veto. It should check if the motion exists and button should change functionality accordingly

**Pull request merged**